acme/acme/client.py

```diff
@@ -285,6 +287,43 @@ def finalize_order(self, orderr: messages.OrderResource, deadline: datetime.date
         self.begin_finalization(orderr)
         return self.poll_finalization(orderr, deadline, fetch_alternative_chains)
 
+    def renewal_time(self, cert_pem: bytes) -> datetime.datetime:
+        """Return an appropriate time to attempt renewal of the certificate.
+
+        If the ACME directory has a "renewalInfo" field, the response will be
+        based on a fetch of the renewal info resource for the certificate
+        (https://www.ietf.org/archive/id/draft-ietf-acme-ari-08.html).
+
+        If there is no "renewalInfo" field, this function will fall back to
+        reasonable defaults based on the certificate lifetime.
+
+        This function may make other network calls in the future (e.g., OCSP
+        or CRL).
+        """
+        cert = x509.load_pem_x509_certificate(cert_pem)
+        ari_path_component = _renewal_info_path_component(cert)
+        try:
+            renewal_info_base_url = self.directory['renewalInfo']
+        except KeyError:
+            not_before = cert.not_valid_before_utc
+            lifetime = cert.not_valid_after_utc - not_before
+            if lifetime.total_seconds() < 10 * 86400:
+                return not_before + lifetime / 2
+            else:
+                return not_before + lifetime * 2 / 3
+        ari_url = renewal_info_base_url + '/' + ari_path_component
+        resp = self.net.get(ari_url, content_type='application/json').json()
+        renewal_info = messages.RenewalInfo.from_json(resp)
```

Comment:
client will crash if this request somehow fail or return error.
ex: when user change to use different CA that support ARI, so while new CA have renewalInfo in directory but it doesn't know about this cert, replying malformed 404 on this request.